man/dracut-catimages.8.adoc
DRACUT-CATIMAGES(8)
===================
:doctype: manpage
:man source:   dracut
:man manual:   dracut

NAME
----
dracut-catimages - creates initial ramdisk image by concatenating images

SYNOPSIS
--------
**dracut-catimages** [_OPTION_...] _<initramfs base image>_ [_<image>_...]

DESCRIPTION
-----------
dracut-catimages creates an initial ramdisk image by concatenating several
images from the command line and /var/lib/dracut/images/*.img

OPTIONS
-------
**-f, --force**::
    overwrite existing initramfs file.

**-i, --imagedir**::
    Directory with additional images to add (default: /var/lib/dracut/images/)

**-o, --overlaydir**::
    Overlay directory, which contains additional files that will be used to
    create an additional image

**--nooverlay**:: Do not use the overlay directory

**--noimagedir**:: Do not use the additional image directory

**-h, --help**:: display help text and exit.

**--debug**:: output debug information of the build process

**-v, --verbose**:: verbose output during the build process

FILES
-----
_/var/lib/dracut/images/*.img_::
    images to work with

AUTHORS
-------
Harald Hoyer

AVAILABILITY
------------
The dracut-catimages command is part of the dracut package and is available from
link:$$https://github.com/dracut-ng/dracut$$[https://github.com/dracut-ng/dracut]

SEE ALSO
--------
man:dracut[8]


NEWS.md
[Rendered view](https://github.com/dracut-ng/dracut/blob/main/NEWS.md)

dracut-ng-111
=============

#### Performance

*   export DRACUT_ARCH globally ([c50c04b0](https://github.com/dracut-ng/dracut/commit/c50c04b00320e170ed77afac19c9753258fe398d))
* **systemd-sysext:**  do not attempt to install extension metadata ([25ffc0f0](https://github.com/dracut-ng/dracut/commit/25ffc0f0da253e49a68f3d5b85e8854a292483c2))

#### Features

*   derive systemd generator name dynamically from script ([6ff36870](https://github.com/dracut-ng/dracut/commit/6ff368706c06bb8608151d12ceee50c56f6bd733))
* **crypt-lib:**  introduce crypt-lib dracut module ([8a82565e](https://github.com/dracut-ng/dracut/commit/8a82565e364b915fa9578e4658312e9c3389de3f))
* **dmsquash-live:**  add busybox blkid compatibility ([7c0298a7](https://github.com/dracut-ng/dracut/commit/7c0298a779368e90c75e690c490415860483faad))
* **dracut:**
  *  add module to add fw files from DT firmware-name properties ([725fe30b](https://github.com/dracut-ng/dracut/commit/725fe30be562e9c5567137f1351ae8c5af0ed422))
  *  add a DRACUT_EXTRA_ARGS environment variable ([99738e52](https://github.com/dracut-ng/dracut/commit/99738e526b7e17ebc0e18bbf51037cb4d0d82304))
* **dracut-catimages.sh:**  move images directory from /boot to /var/lib ([6bfd026e](https://github.com/dracut-ng/dracut/commit/6bfd026ee164d6b2a5ba4a2b91ae9d0984e4a2f7))
* **overlayfs:**  support tmpfs size parameter in rd.overlay ([9740a99a](https://github.com/dracut-ng/dracut/commit/9740a99a5e1164db7d64c8bd602427f2239ad5e3))
* **overlayfs-crypt:**  add new encrypted persistent overlay support ([c74f8ac2](https://github.com/dracut-ng/dracut/commit/c74f8ac20500211edd02f4fa9acb17c07c1b3a38))
* **systemd-sysext:**  include systemd-{sys,conf}ext-sysroot services ([205415b5](https://github.com/dracut-ng/dracut/commit/205415b52f2e413780a86bc33784ee99f8ace4cf))
* **systemd-sysusers-service:**  systemd-sysusers.service module ([e0a7c9e2](https://github.com/dracut-ng/dracut/commit/e0a7c9e2f1bacd283419320166e5630503d8fc68))

#### Bug Fixes

*   disable hostonly_cmdline for Gentoo by default ([6c772faa](https://github.com/dracut-ng/dracut/commit/6c772faab608c078a1aa3ff3fa509679629485ff))
*   explicitly set and document Gentoo defaults ([86dae808](https://github.com/dracut-ng/dracut/commit/86dae8081027e3461a49999d63938561eb2ed662))
*   avoid arguments for dot commands ([045e9e30](https://github.com/dracut-ng/dracut/commit/045e9e306f6d8a9359c1e9265200d9d4b6c85ec3))
*   set DRACUT_TMPDIR before using it in dlog_init() ([56ba87ad](https://github.com/dracut-ng/dracut/commit/56ba87ad233adc8f6ce4b13218efd1e18c297035))
* **SYSTEMD-IMPORT:**  remount /sysroot with dev and suid flags ([32d12606](https://github.com/dracut-ng/dracut/commit/32d1260667d0bea64573bf23ffa8f5c23c631d2a))
* **base:**  move initrd.target.wants symlink creation to wait_for_dev ([d971230a](https://github.com/dracut-ng/dracut/commit/d971230a5a8c0079ac342d4b5e03bf8b5496e6b1))
* **crypt:**
  *  use systemd generator for setting the timeout ([951be8ce](https://github.com/dracut-ng/dracut/commit/951be8ce50d08b6dc8cf6b474c2bfb632048df9a))
  *  use -d instead of -f to check for $NEWROOT/proc directory ([57edbca1](https://github.com/dracut-ng/dracut/commit/57edbca16f6a819c2f12058a8f1ea609e2a9c7ef))
  *  honor timeout setting when using UUID, LABEL, etc ([89d5f2cb](https://github.com/dracut-ng/dracut/commit/89d5f2cb7e9d31ac7d8ab1ba6b5757cc2c9ec403))
* **dbus:**  remove After/Requires from dbus service/socket ([6102479e](https://github.com/dracut-ng/dracut/commit/6102479e83f1c3f3e57f0a46a90494489dfeaa9c))
* **dmsquash-live:**
  *  correct det_img_fs implementation ([f3d81b8d](https://github.com/dracut-ng/dracut/commit/f3d81b8d968762896d60e2061a6c2b7d400fc2d0))
  *  load kernel module before mount ([9ca2cec1](https://github.com/dracut-ng/dracut/commit/9ca2cec17ab21049b81ad411fb18604faddd58b3))
* **dracut:**
  *  do not record parameters that do not change the initrd ([44a15508](https://github.com/dracut-ng/dracut/commit/44a15508c0bf77e836c7745c1ec3be48b2d6559b))
  *  remove leading space from recorded arguments ([0a62cc4a](https://github.com/dracut-ng/dracut/commit/0a62cc4a27568e1b19b4e9c6d86bbc9cf0693884))
  *  inst_libdir_file() does not have a -o option ([a8b2a995](https://github.com/dracut-ng/dracut/commit/a8b2a995272d1e63494c1b8b74a908db5137a1b6))
  *  determine hostonly_cmdline after hostonly setting ([6cfafda4](https://github.com/dracut-ng/dracut/commit/6cfafda44a70c9907e8af0a11dba95c36896d369))
  *  properly detect kernel version with --sysroot ([d19f65d1](https://github.com/dracut-ng/dracut/commit/d19f65d17e0b8d5cc7595339b3a83092e8d760ed))
  *  enable hostonly_cmdline in hostonly mode again ([04af3c09](https://github.com/dracut-ng/dracut/commit/04af3c097797d102934bc7ec666ed8eef00af9bb))
  *  remove unused dracut_no_switch_root ([f61865ce](https://github.com/dracut-ng/dracut/commit/f61865ce5288c4876d5d32ad2e2e8ab0ab4bf3f0))
  *  remove udevrulesconfdir variable ([d867f00d](https://github.com/dracut-ng/dracut/commit/d867f00d87b7ccc2738901a73f3d7bc20ed30989))
  *  remove unused host_fs_all ([8ef39d08](https://github.com/dracut-ng/dracut/commit/8ef39d0802daade93d7f578cfd6bb3bfc92a139c))
* **dracut-catimages.sh:**
  *  do not create overlay image without image directory ([66336959](https://github.com/dracut-ng/dracut/commit/66336959b8e225499d1868888ec31fa881b8f4b5))
  *  do not call shift for options without arguments ([e46ac221](https://github.com/dracut-ng/dracut/commit/e46ac221e2ac7d5a3f60e0f66173ea585ef98710))
  *  log proper missing images directory ([1d78fd0e](https://github.com/dracut-ng/dracut/commit/1d78fd0e0a014addd7c6fe56c42d1c044628fa46))
* **dracut.sh:**  use printf instead of echo in mark_hostonly ([002f0901](https://github.com/dracut-ng/dracut/commit/002f09013ef3c9c98b7ec89ed5e35a7fa8f161e6))

dracut-catimages.sh
#!/bin/bash --norc
#
# Copyright 2009 Red Hat, Inc.  All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

dinfo() {
    [[ $beverbose ]] && echo "$@" >&2
}

derror() {
    echo "Error: $*" >&2
}

usage() {
    #                                                   80x25 linebreak here ^
    cat << EOF
Usage: $0 [OPTION]... <initramfs> <base image> [<image>...]
Creates initial ramdisk image by concatenating several images from the command
line and /var/lib/dracut/images/

  -f, --force           Overwrite existing initramfs file.
  -i, --imagedir        Directory with additional images to add
                        (default: /var/lib/dracut/images/)
  -o, --overlaydir      Overlay directory, which contains files that
                        will be used to create an additional image
  --nooverlay           Do not use the overlay directory
  --noimagedir          Do not use the additional image directory
  -h, --help            This message
  --debug               Output debug information of the build process
  -v, --verbose         Verbose output during the build process
EOF
}

imagedir=/var/lib/dracut/images
overlay=/var/lib/dracut/overlay

while (($# > 0)); do
    case $1 in
        -f | --force) force=yes ;;
        -i | --imagedir)
            imagedir=$2
            shift
            ;;
        -o | --overlaydir)
            overlay=$2
            shift
            ;;
        --nooverlay)
            no_overlay=yes
            ;;
        --noimagedir)
            no_imagedir=yes
            ;;
        -h | --help)
            usage
            exit 1
            ;;
        --debug) export debug="yes" ;;
        -v | --verbose) beverbose="yes" ;;
        -*)
            printf "\nUnknown option: %s\n\n" "$1" >&2
            usage
            exit 1
            ;;
        *) break ;;
    esac
    shift
done

outfile=$1
shift

if [[ -z $outfile ]]; then
    derror "No output file specified."
    usage
    exit 1
fi

baseimage=$1
shift

if [[ -z $baseimage ]]; then
    derror "No base image specified."
    usage
    exit 1
fi

if [[ -f $outfile && ! $force ]]; then
    derror "Will not override existing initramfs ($outfile) without --force"
    exit 1
fi

if [[ ! $no_imagedir && ! -d $imagedir ]]; then
    derror "Image directory $imagedir is not a directory"
    exit 1
fi

if [[ ! $no_overlay && ! -d $overlay ]]; then
    derror "Overlay $overlay is not a directory"
    exit 1
fi

if [[ ! $no_overlay ]]; then
    if [[ ! $no_imagedir ]]; then
        ofile="$imagedir/90-overlay.img"
        dinfo "Creating image $ofile from directory $overlay"
        type pigz &> /dev/null && gzip=pigz || gzip=gzip
        (
            cd "$overlay" || return 1
            find . | cpio --quiet -H newc -o | $gzip -9 > "$ofile"
        )
    else
        dinfo "Overlay image will not be created without the additional image directory"
    fi
fi

if [[ ! $no_imagedir ]]; then
    for i in "$imagedir/"*.img; do
        [[ -f $i ]] && images+=("$i")
    done
fi

images+=("$@")

dinfo "Using base image $baseimage"
cat -- "$baseimage" > "$outfile"

for i in "${images[@]}"; do
    dinfo "Appending $i"
    cat -- "$i" >> "$outfile"
done

dinfo "Created $outfile"

exit 0
